File: fundamentos/desafio-js/src/questions/calc-total-of-highest-acc-per-agency.ts
import { IAccount } from "../i-account";

/** Considere o cliente com o maior saldo em cada agência (caso haja mais de um cliente com o maior saldo, escolha apenas um). O valor total desses saldos é: */
export default function calcTotalOfHigestAccPerAgency(...accounts: Array<IAccount>) {
  const agencies = new Map<number, IAccount>();

  for (const acc of accounts) {
    if (agencies.has(acc.agencia)) {
      if ((agencies.get(acc.agencia)?.balance || 0) < acc.balance) {
        agencies.set(acc.agencia, acc);
      }
    } else {
      agencies.set(acc.agencia, acc);
    }
  }

  const sun = Array.from(agencies.values())
    .reduce((prev, item) => ({...prev, balance: prev.balance + item.balance}), { balance: 0 });

  console.info(`Considere o cliente com o maior saldo em cada agência (caso haja mais de um cliente com o maior saldo, escolha apenas um). O valor total desses saldos é: $${sun.balance}`);

}

File: fundamentos/desafio-dart/lib/questions/calc_total_of_highest_acc_per_agency.dart
import 'package:desafio_dart/domains/account.dart';

/// Considere o cliente com o maior saldo em cada agência (caso haja mais de um cliente com o maior saldo,
/// escolha apenas um). O valor total desses saldos é:
void calcTotalOfHigestAccPerAgency(List<Account> accounts) {
  var agencies = Map<int, Account>();

  for (var acc in accounts) {
    if (agencies.containsKey(acc.agencia)) {
      if ((agencies[acc.agencia].balance ?? 0) < acc.balance) {
        agencies[acc.agencia] = acc;
      }
    } else {
      agencies[acc.agencia] = acc;
    }
  }

  var sum = agencies.values.reduce((prev, item) =>
      item.copyWith(balance: (prev?.balance ?? 0) + item.balance));

  print(
      'Considere o cliente com o maior saldo em cada agência (caso haja mais de um cliente com o maior saldo, escolha apenas um). O valor total desses saldos é: \$${sum.balance}');
}
